app/src/main/java/org/schabi/newpipe/player/Player.java

```diff
@@ -3355,22 +3497,232 @@ private void updateQueueTime(final int currentTime) {
+     * </p>
+     */
+    private void buildVideoQualityMenuIfNeeded() {
+        if (currentMetadata == null || qualityPopupMenu == null || qualityPopupMenuBuilt) {
             return;
         }
-        qualityPopupMenu.getMenu().removeGroup(POPUP_MENU_ID_QUALITY);
 
-        for (int i = 0; i < availableStreams.size(); i++) {
-            final VideoStream videoStream = availableStreams.get(i);
-            qualityPopupMenu.getMenu().add(POPUP_MENU_ID_QUALITY, i, Menu.NONE, MediaFormat
-                    .getNameById(videoStream.getFormatId()) + " " + videoStream.resolution);
-        }
-        if (getSelectedVideoStream() != null) {
-            binding.qualityTextView.setText(getSelectedVideoStream().resolution);
+        final Menu menu = qualityPopupMenu.getMenu();
+        menu.removeGroup(POPUP_MENU_ID_QUALITY);
+
+        for (int i = 0; i < availableVideoStreams.size(); i++) {
+            final VideoStream videoStream = availableVideoStreams.get(i);
+            menu.add(POPUP_MENU_ID_QUALITY, i, Menu.NONE, MediaFormat.getNameById(
+                    videoStream.getFormatId()) + " " + videoStream.resolution);
         }
+
         qualityPopupMenu.setOnMenuItemClickListener(this);
         qualityPopupMenu.setOnDismissListener(this);
+        qualityPopupMenuBuilt = true;
+    }
+
+    /**
+     * Build the quality menu for livestreams, if needed.
+     *
+     * <p>
+     * This method loops into available streams in the video renderer index and add them to the
+     * quality selector.
+     * </p>
+     *
+     * <p>
+     * Only the resolution is shown, with the frame rate when its value is more than 30 because it
+     * seems that there is no way to get the container format/mime type.
+     * </p>
+     *
+     * @param videoTrackGroupArray the video {@link TrackGroupArray}, which cannot be null
+     * @param videoRendererIndex   the index of the video renderer
+     */
+    private void buildLiveStreamQualityMenuIfNeeded(
+            @NonNull final TrackGroupArray videoTrackGroupArray,
+            final int videoRendererIndex) {
+        if (currentMetadata == null || qualityPopupMenu == null || qualityPopupMenuBuilt) {
+            return;
+        }
+
+        int videoFormatsCount = 0;
+        for (int i = 0; i < videoTrackGroupArray.length; i++) {
+            videoFormatsCount += videoTrackGroupArray.get(i).length;
+        }
+
+        final Menu menu = qualityPopupMenu.getMenu();
+        if (videoFormatsCount + 1 == menu.size()) {
+            return;
+        }
+
+        menu.removeGroup(POPUP_MENU_ID_QUALITY);
+        menu.add(POPUP_MENU_ID_QUALITY, 0, Menu.NONE,
+                context.getString(R.string.auto_quality));
+
+        int menuItemId = 1;
+        for (int i = 0; i < videoTrackGroupArray.length; i++) {
+            final TrackGroup videoTrackGroup = videoTrackGroupArray.get(i);
+                final int videoTrackGroupLength = videoTrackGroup.length;
```

Comment:
Do not create this variable, just use `videoTrackGroup.length` each time
```suggestion
```

Reply:
Why? Is not better to store the length instead of getting it at each loop tour?

(I fixed the indentation by the way.)

Reply:
You are not "getting" it. You are reading it directly. If you were calling a `getLength()` method then yes, it would be slightly faster to store it (though probably an unneeded optimization). But since there is no method call but just reading from a class field, the performance should be exactly the same.

Reply:
Ok, no problem. I will change this.